### ISSUE TITLE
/usr/bin/ld: cannot find -lstdc++

### ISSUE THREAD

**Issue by andyroberts007** [Ultimaker/libArcus#30]
Trying to build on rhel 7
did a yum update

installed protobuf
installed Python3.4
installed sip-4.18
installed cmake-3.5.2

git cloned libArcus
created build dir
ran cmake .. OK

make fails as follows:
make
Scanning dependencies of target Arcus
[  4%] Building CXX object CMakeFiles/Arcus.dir/src/Socket.cpp.o
[  9%] Building CXX object CMakeFiles/Arcus.dir/src/SocketListener.cpp.o
[ 13%] Building CXX object CMakeFiles/Arcus.dir/src/MessageTypeStore.cpp.o
[ 18%] Building CXX object CMakeFiles/Arcus.dir/src/PlatformSocket.cpp.o
[ 22%] Building CXX object CMakeFiles/Arcus.dir/src/Error.cpp.o
[ 27%] Linking CXX shared library libArcus.so
/usr/bin/ld: cannot find -lstdc++
collect2: error: ld returned 1 exit status
make[2]: *** [libArcus.so.1.0.0] Error 1
make[1]: *** [CMakeFiles/Arcus.dir/all] Error 2
make: *** [all] Error 2

but I seemingly have libstdc++

yum search libstdc++
Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-
              : manager
Repo rhel-7-workstation-rpms forced skip_if_unavailable=True due to: /etc/pki/entitlement/4886347605660102700-key.pem
============================ N/S matched: libstdc++ ============================
libstdc++.i686 : GNU Standard C++ Library
libstdc++.x86_64 : GNU Standard C++ Library
libstdc++-devel.i686 : Header files and libraries for C++ development
libstdc++-devel.x86_64 : Header files and libraries for C++ development
libstdc++-docs.x86_64 : Documentation for the GNU standard C++ library

Any ideas where I am going wrong?

Thanks - Andy Roberts

**Comment by andyroberts007** [comment]
a bit more info on the problem

 yum provides \*/libstdc++.so
Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-manager
rhel-7-workstation-rpms/7Workstation/x86_64/filelists_db                                                   |  11 MB  00:00:25     
compat-gcc-44-c++-4.4.7-8.el7.x86_64 : C++ support for compatibility compiler
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.4.7/32/libstdc++.so
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.4.7/libstdc++.so



gcc-c++-4.8.2-16.el7.x86_64 : C++ support for GCC
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/32/libstdc++.so
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libstdc++.so



gcc-c++-4.8.2-16.2.el7_0.x86_64 : C++ support for GCC
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/32/libstdc++.so
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libstdc++.so



gcc-c++-4.8.3-9.el7.x86_64 : C++ support for GCC
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/32/libstdc++.so
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libstdc++.so



gcc-c++-4.8.5-4.el7.x86_64 : C++ support for GCC
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/32/libstdc++.so
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libstdc++.so



libstdc++-devel-4.8.2-16.el7.i686 : Header files and libraries for C++ development
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/i686-redhat-linux/4.8.2/libstdc++.so



libstdc++-devel-4.8.2-16.2.el7_0.i686 : Header files and libraries for C++ development
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/i686-redhat-linux/4.8.2/libstdc++.so



libstdc++-devel-4.8.3-9.el7.i686 : Header files and libraries for C++ development
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/i686-redhat-linux/4.8.2/libstdc++.so



libstdc++-devel-4.8.5-4.el7.i686 : Header files and libraries for C++ development
Repo        : rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/i686-redhat-linux/4.8.2/libstdc++.so



gcc-c++-4.8.5-4.el7.x86_64 : C++ support for GCC
Repo        : @rhel-7-workstation-rpms
Matched from:
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/32/libstdc++.so
Filename    : /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libstdc++.so

**Comment by Ghostkeeper** [comment]
Try this:

    sudo yum install libstdc++-static

**Comment by andyroberts007** [comment]
I've tried that but it's not available on rhel-7

yum install libstdc++-static
Loaded plugins: langpacks, product-id, search-disabled-repos, subscription-manager
No package libstdc++-static available.
Error: Nothing to do

I tried the following:
gcc -m64 -Xlinker --verbose 2>/dev/null | grep SEARCH | sed 's/SEARCH_DIR("=\?\([^"]\+\)"); */\1\n/g' | grep -vE '^$'
/usr/x86_64-redhat-linux/lib64
/usr/local/lib64
/lib64
/usr/lib64
/usr/x86_64-redhat-linux/lib
/usr/lib64
/usr/local/lib
/lib
/usr/lib

From this I noticed that libstdc++.so was not on any of these paths so I did a hack link

ln -s /usr/lib/gcc/x86_64-redhat-linux/4.8.5/libstdc++.so /usr/lib/libstdc++.so

this didn't work either

cmake ..
-- The C compiler identification is GNU 4.8.5
-- The CXX compiler identification is GNU 4.8.5
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found Protobuf: /usr/local/lib/libprotobuf.so (Required is at least version "3.0.0") 
-- Found PythonInterp: /usr/local/bin/python3.4 (found suitable version "3.4.4", minimum required is "3.4.0") 
-- Found PythonLibs: /usr/local/lib/libpython3.4m.a (found suitable version "3.4.4", minimum required is "3.4.0") 
-- Found SIP version: 4.18
-- Configuring done
-- Generating done
-- Build files have been written to: /home/andyroberts/libArcus/build
[andyroberts@localhost build]$ make
Scanning dependencies of target Arcus
[  4%] Building CXX object CMakeFiles/Arcus.dir/src/Socket.cpp.o
[  9%] Building CXX object CMakeFiles/Arcus.dir/src/SocketListener.cpp.o
[ 13%] Building CXX object CMakeFiles/Arcus.dir/src/MessageTypeStore.cpp.o
[ 18%] Building CXX object CMakeFiles/Arcus.dir/src/PlatformSocket.cpp.o
[ 22%] Building CXX object CMakeFiles/Arcus.dir/src/Error.cpp.o
[ 27%] Linking CXX shared library libArcus.so
/usr/bin/ld: cannot find -lstdc++
collect2: error: ld returned 1 exit status
make[2]: *** [libArcus.so.1.0.0] Error 1
make[1]: *** [CMakeFiles/Arcus.dir/all] Error 2
make: *** [all] Error 2


Still digging...

**Comment by andyroberts007** [comment]
but with the hack link, I get this with ld

ld -lstdc++
ld: warning: cannot find entry symbol _start; not setting start address

Not sure if I need to have the static version of this installed - and if so how to do it...

**Comment by andyroberts007** [comment]
some progress - did the following and the link now works

rpm -ivh libstdc++-static-4.8.5-4.el7.x86_64.rpm

**Comment by andyroberts007** [comment]
but now get this error

make
Scanning dependencies of target Arcus
[  4%] Building CXX object CMakeFiles/Arcus.dir/src/Socket.cpp.o
[  9%] Building CXX object CMakeFiles/Arcus.dir/src/SocketListener.cpp.o
[ 13%] Building CXX object CMakeFiles/Arcus.dir/src/MessageTypeStore.cpp.o
[ 18%] Building CXX object CMakeFiles/Arcus.dir/src/PlatformSocket.cpp.o
[ 22%] Building CXX object CMakeFiles/Arcus.dir/src/Error.cpp.o
[ 27%] Linking CXX shared library libArcus.so
[ 27%] Built target Arcus
[ 31%] Generating python/sipArcuspart0.cpp, python/sipArcuspart1.cpp, python/sipArcuspart2.cpp, python/sipArcuspart3.cpp, python/sipArcuspart4.cpp, python/sipArcuspart5.cpp, python/sipArcuspart6.cpp, python/sipArcuspart7.cpp

Scanning dependencies of target python_module_Arcus
[ 36%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart0.cpp.o
[ 40%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart1.cpp.o
[ 45%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart2.cpp.o
[ 50%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart3.cpp.o
[ 54%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart4.cpp.o
[ 59%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart5.cpp.o
[ 63%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart6.cpp.o
[ 68%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/sipArcuspart7.cpp.o
[ 72%] Building CXX object CMakeFiles/python_module_Arcus.dir/python/PythonMessage.cpp.o
[ 77%] Linking CXX shared library Arcus.so
/usr/bin/ld: /usr/local/lib/libpython3.4m.a(abstract.o): relocation R_X86_64_32S against `_Py_NotImplementedStruct' can not be used when making a shared object; recompile with -fPIC
/usr/local/lib/libpython3.4m.a: could not read symbols: Bad value
collect2: error: ld returned 1 exit status
make[2]: *** [Arcus.so] Error 1
make[1]: *** [CMakeFiles/python_module_Arcus.dir/all] Error 2
make: *** [all] Error 2
[andyroberts@localhost build]$ make -fPIC
make: PIC: No such file or directory
make: *** No rule to make target `PIC'.  Stop.

**Comment by andyroberts007** [comment]
sort of resolved

So I went back to my Python3.4 install and rebuilt using --enable-shared
Then I rebuild cmake and tried building libArcus.  This gave an error that only python 1 could be found.
So I went back to Python3.4 without --enable-shared and everything compiles now.

The original problem of this thread was solved by doing the following:
rpm -ivh libstdc++-static-4.8.5-4.el7.x86_64.rpm
